src/openvpn/tun2tap.c

```diff
@@ -0,0 +1,196 @@
+        if ( BLEN(&c->c2.buf) >= sizeof(struct openvpn_iphdr)){
+            struct openvpn_ethhdr hdr = {0};
+            struct openvpn_iphdr *ip_hdr = BPTR(&c->c2.buf);
+            struct openvpn_ipv6hdr *ipv6_hdr = BPTR(&c->c2.buf);
+            int v = OPENVPN_IPH_GET_VER(ip_hdr->version_len);
+            memcpy(hdr.dest, c->c1.tuntap->remote_mac_addr, OPENVPN_ETH_ALEN);
+            memcpy(hdr.source, c->options.lladdr_v, OPENVPN_ETH_ALEN);
+            
+            if (0 == memcmp(hdr.dest, "\x00\x00\x00\x00\x00\x00", sizeof("\x00\x00\x00\x00\x00\x00") - 1)){
+                // build request
+                // overwirte arp to buf
+                struct openvpn_arp arp = { 0 };
+                arp.mac_addr_type = htons(0x0001);
+                arp.proto_addr_type = htons(0x0800);
+                arp.mac_addr_size = 0x06;
+                arp.proto_addr_size = 0x04;
+                arp.arp_command = htons(ARP_REQUEST);
+                memcpy(arp.mac_src, hdr.source, OPENVPN_ETH_ALEN);
+                memcpy(arp.mac_dest, "\x00\x00\x00\x00\x00\x00", OPENVPN_ETH_ALEN);
+                memcpy(hdr.dest, "\xff\xff\xff\xff\xff\xff", OPENVPN_ETH_ALEN);
+                hdr.proto = htons(OPENVPN_ETH_P_ARP);
+                switch (v)
+                {
+                case 4:
+                    arp.ip_src = ip_hdr->saddr;
+                    arp.ip_dest = ip_hdr->daddr;
+                    break;
+                case 6: 
+                    // hdr.proto = htons(OPENVPN_ETH_P_IPV6);
+                    // memcpy(&arp.ip_src, &ipv6_hdr->saddr, sizeof(ipv6_hdr->saddr));
+                    // memcpy(&arp.ip_dest, &ipv6_hdr->daddr, sizeof(ipv6_hdr->daddr));
+                default:
+                    break;
+                }
+                buf_clear(&c->c2.buf);
+                buf_write(&c->c2.buf, &hdr, sizeof(hdr));
+                buf_write(&c->c2.buf, &arp, sizeof(arp));
+            } else {
+                switch (v)
+                {
+                case 4:
+                    hdr.proto = htons(OPENVPN_ETH_P_IPV4);
+                    break;
+                case 6: 
+                    hdr.proto = htons(OPENVPN_ETH_P_IPV6);
+                default:
+                    break;
+                }
+                ASSERT(buf_write_prepend(&c->c2.buf, &hdr, sizeof(hdr)));
+            }
+            dmsg(D_TUN2TAP, "dest addr is: %02x:%02x:%02x:%02x:%02x:%02x"
+                , hdr.dest[0]
+                , hdr.dest[1]
+                , hdr.dest[2]
+                , hdr.dest[3]
+                , hdr.dest[4]
+                , hdr.dest[5]
+            );
+            dmsg(D_TUN2TAP, "source addr is: %02x:%02x:%02x:%02x:%02x:%02x"
+                , hdr.source[0]
+                , hdr.source[1]
+                , hdr.source[2]
+                , hdr.source[3]
+                , hdr.source[4]
+                , hdr.source[5]
+            );
+        }
+    } else if (TUN2TAP_FLAG_DECAP == flag){
```

Comment:
It looks like this function is really two function separated by an if. Make it two functions!